### ISSUE TITLE
v8.9.0 input processing and/or transition issues

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#6601]
Issue overview
--------------
- [x] For Boilers and Chillers, the Flow Mode choices were changed back in v8.0, but the getinputs still accepted "VariableFlow" with a warning. Since "VariableFlow" is not a valid IDD key, the v8.9 input processor kicks this out. Transition to v8.9 needs to clean up any remaining "VariableFlow" and change them to "LeavingSetpointModulated". Then for the next release, the getinputs need to be modified to get rid of the [blocks](https://github.com/NREL/EnergyPlus/blob/c8f9fbba47a5425fbb4220285f70d0017f6f65b4/src/EnergyPlus/Boilers.cc#L492-L496) that warn and change this key.
[Hospital CHP-storage v8-5.idf.txt](https://github.com/NREL/EnergyPlus/files/1857681/Hospital.CHP-storage.v8-5.idf.txt)
Helpdesk ticket number 12516.
*Addressed in #7532*

See comments below for additional issues.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by mjwitte** [comment]
- [x]   EnergyManagementSystem:Program does not accept a blank line in the middle of the object. Either need to transition these away or modify input processor to allow this.
For example:
```
  EnergyManagementSystem:Program,
    ER_Main,                 !- Name
    IF ER_Humidifier_Status > 0,  !- Program Line 1
    SET ER_ExtraElecHeatC_Status = 1,  !- Program Line 2
    SET ER_ExtraElecHeatC_SP = ER_AfterHumidifier_Temp + 1.4,  !- <none>
    ELSE,                    !- <none>
    SET ER_ExtraElecHeatC_Status = 0,  !- <none>
    SET ER_ExtraElecHeatC_SP = NULL,  !- <none>
    ENDIF,                   !- <none>
    IF T_OA < 10,            !- <none>
    ,                        !- <none>
    SET HeatGain = 0 * (ER_FanDesignMass/1.2) *2118,  !- <none>
    SET FlowRate = (ER_FanMassFlow/1.2)*2118,  !- <none>
    SET ER_PreheatDeltaT = HeatGain/(1.08*(FLOWRATE+0.000001)),  !- <none>
    SET ER_ExtraWaterHeatC_Status = 1,  !- <none>
    SET ER_ExtraWaterHeatC_SP = ER_AfterElecHeatC_Temp + ER_PreheatDeltaT,  !- <none>
    ELSE,                    !- <none>
    SET ER_ExtraWaterHeatC_Status = 0,  !- <none>
    SET ER_ExtraWaterHeatC_SP = NULL,  !- <none>
    ENDIF;                   !- <none>

   ** Severe  ** <root>[EnergyManagementSystem:Program][ER_Main][lines][8] - Missing required property 'program_line'.
```
JSON out looks like this:
```
                {
                    "program_line": "IF T_OA < 10"
                },
                {},
                {
                    "program_line": "SET HeatGain = 0 * (ER_FanDesignMass/1.2) *2118"
                },

```
Same defect file, [Hospital CHP-storage v8-5.idf.txt](https://github.com/NREL/EnergyPlus/files/1857681/Hospital.CHP-storage.v8-5.idf.txt)
Helpdesk ticket number 12516.

*Addressed by #6832*

**Comment by mjwitte** [comment]
- [x] Fuel types allow many synonyms in the get inputs, [WaterHeater:Mixed, for example](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/WaterThermalTanks.cc#L2564-L2589) allows "Gas", "NaturalGas" and "Natural Gas" in the getinput, but the only currently valid key in the idd (for gas) is "NaturalGas". Anything else fails in the input processor.  Need to transition these all away and look for others in the code.

Same defect file [Hospital CHP-storage v8-5.idf.txt](https://github.com/NREL/EnergyPlus/files/1857681/Hospital.CHP-storage.v8-5.idf.txt)
Helpdesk ticket number 12516.
*Partially addressed in #7532, remainder will be addressed in #7523*

**Comment by mjwitte** [comment]
- [x] HeatBalanceAlgorithm [synonyms](https://github.com/NREL/EnergyPlus/blob/c8f9fbba47a5425fbb4220285f70d0017f6f65b4/src/EnergyPlus/HeatBalanceManager.cc#L954-L976) need to be transitioned away, e.g.
```
			if ( ( SELECT_CASE_var == "CONDUCTIONTRANSFERFUNCTION" ) || ( SELECT_CASE_var == "DEFAULT" ) || ( SELECT_CASE_var == "CTF" ) ) {
```
No test file.
*Addressed in #7532*

**Comment by mjwitte** [comment]
- [x] ZoneAirHeatBalanceAlgorithm [synonyms](https://github.com/NREL/EnergyPlus/blob/c8f9fbba47a5425fbb4220285f70d0017f6f65b4/src/EnergyPlus/HeatBalanceManager.cc#L1046) need to be transitioned away.
```
				if ( ( SELECT_CASE_var == "3RDORDERBACKWARDDIFFERENCE" ) || ( SELECT_CASE_var == "THIRDORDERBACKWARDDIFFERENCE" ) ) {
```
No test file.
*Addressed in #7532*

**Comment by mjwitte** [comment]
- [x] Zone "Outside Convection Algorithm" [synonyms](https://github.com/NREL/EnergyPlus/blob/c8f9fbba47a5425fbb4220285f70d0017f6f65b4/src/EnergyPlus/HeatBalanceManager.cc#L4707)
```
				} else if ( ( SELECT_CASE_var == "DOE2" ) || ( SELECT_CASE_var == "DOE-2" ) ) {

```
*Addressed in #7532*

**Comment by mjwitte** [comment]
- [x] Generator:WindTurbine "Rotor Type" [synonyms](https://github.com/NREL/EnergyPlus/blob/c8f9fbba47a5425fbb4220285f70d0017f6f65b4/src/EnergyPlus/WindTurbine.cc#L312-L334), e.g.
```
			} else if ( ( SELECT_CASE_var == "VERTICALAXISWINDTURBINE" ) || ( SELECT_CASE_var == "VAWT" ) ) {
```
*Addressed in #7532*

**Comment by mjwitte** [comment]
- [x] @mitchute The v8.9 transition rules for GLHE are not described in the [rules file](https://github.com/NREL/EnergyPlus/blob/develop/src/Transition/InputRulesFiles/Rules8-8-0-to-8-9-0.md)
*Addressed in #7532*

**Comment by SchildCode** [comment]
I fixed this problem on my copy of E+ v.9.1.0 by reverting to the v.8.9.0 schema for object "EnergyManagementSystem:Program". It appears that the EnergyManagementSystem:Program object was changed in the schema between versions 8.9.0 and 9.1.0.  The 9.1.0 version requires field "lines" (which is the name of the extended group, but not the name of a field); however the is no field called "lines", it should require extended field "program_line" instead. The change between 8.9.0 and 9.1.0 breaks the established practice that only fields listed in legacy_idd:{fields:[]} or legacy_idd:{extensibles[]} can be required, not extension group names. 

### **Original epJSON object schema in v.8.9.0:**
        "EnergyManagementSystem:Program": {
            "extensible_size": 1.0, 
            "name": {
                "note": "no spaces allowed in name", 
                "is_required": true, 
                "type": "string", 
                "reference": [
                    "ErlProgramNames"
                ]
            }, 
            "min_fields": 2.0, 
            "patternProperties": {
                ".*": {
                    "type": "object", 
                    "properties": {
                        "lines": {
                            "items": {
                                "required": [
                                    "program_line"
                                ], 
                                "type": "object", 
                                "properties": {
                                    "program_line": {
                                        "type": "string"
                                    }
                                }
                            }, 
                            "type": "array"
                        }
                    }
                }
            }, 
            "legacy_idd": {
                "numerics": {
                    "fields": []
                }, 
                "field_info": {
                    "name": {
                        "field_type": "a", 
                        "field_name": "Name"
                    }, 
                    "program_line": {
                        "field_type": "a", 
                        "field_name": "Program Line"
                    }
                }, 
                "extensibles": [
                    "program_line"
                ], 
                "extension": "lines", 
                "alphas": {
                    "fields": [
                        "name"
                    ], 
                    "extensions": [
                        "program_line"
                    ]
                }, 
                "fields": [
                    "name"
                ]
            }, 
            "type": "object", 
            "memo": "This input defines an Erl program Each field after the name is a line of EMS Runtime Language"
        }, 

### **Corrupt object schema in v.9.1.0:**
        "EnergyManagementSystem:Program": {
            "extensible_size": 1.0, 
            "name": {
                "note": "no spaces allowed in name", 
                "is_required": true, 
                "type": "string", 
                "reference": [
                    "ErlProgramNames"
                ]
            }, 
            "min_fields": 2.0, 
            "patternProperties": {
                ".*": {
                    "required": [
                        "lines"   <=== OFFENDING LINE
                    ], 
                    "type": "object", 
                    "properties": {
                        "lines": {
                            "minItems": 1, 
                            "items": {
                                "type": "object", 
                                "properties": {
                                    "program_line": {
                                        "type": "string"
                                    }
                                }
                            }, 
                            "type": "array"
                        }
                    }
                }
            }, 
            "legacy_idd": {
                "numerics": {
                    "fields": []
                }, 
                "field_info": {
                    "name": {
                        "field_type": "a", 
                        "field_name": "Name"
                    }, 
                    "program_line": {
                        "field_type": "a", 
                        "field_name": "Program Line"
                    }
                }, 
                "extensibles": [
                    "program_line"
                ], 
                "extension": "lines", 
                "alphas": {
                    "fields": [
                        "name"
                    ], 
                    "extensions": [
                        "program_line"
                    ]
                }, 
                "fields": [
                    "name"
                ]
            }, 
            "type": "object", 
            "memo": "This input defines an Erl program Each field after the name is a line of EMS Runtime Language"
        },

**Comment by mjwitte** [comment]
- [x] GlobalGeometryRules: ULC, CCW, etc.
*Addressed in #7523*

**Comment by mjwitte** [comment]
- [x] Hmmm, transitioned a v810 file that had 0's and 1's for SizingPeriod:DesignDay fields for rain/snow/dst. These should be Yes/No - not sure when that changed and why it didn't transition.  See 4397 defect file.
*Addressed in #7532*

- [x] Same file   Output:Constructions,Construction;  should be   Output:Constructions,Constructions; same for Material/s
*Addressed in #7532*

**Comment by mjwitte** [comment]
@mbadams5 @Myoldmopar  I've been working on cleaning out undocumented input synonyms from the EnergyPlus code and adding transition rules to CreateNewIDFUsingRulesV8_9_0.f90, naively thinking the key choices in the IDD were enforced by the input processor as of v8.9. 

But now I see that [modify_schema.py](https://github.com/NREL/EnergyPlus/blob/60ed7207c2f3058fe8dc400a1aa741149094bebe/scripts/dev/generate_epJSON_schema/modify_schema.py#L267-L297) modifies some objects in the schema by adding new key choices to some objects or removing the key choice list altogether (allowing any string to pass to the getinput function). So this means old strings like "ULC" for UpperLeftCorner are still accepted (and are still lurking in some testfiles). So those transition rules need to be in CreateNewIDFUsingRulesV9_2_0.f90 (or more likely now 9_3_0). But some of the items listed in this issue did not get special treatment and do indeed need to be transitioned away back in CreateNewIDFUsingRulesV8_9_0.f90.

@mbadams5 Before I continue down this path do you agree that the goal should be to eliminate as many of these schema modifications as possible and use transition to standardize any idfs to match the IDD key choices? Is there some reason to keep any of the old synonyms active?

@Myoldmopar Needless to say this will be for post v9.2 release - too complicated to cram in the transition rules now.

**Comment by mbadams5** [comment]
@mjwitte Yes, the goal should be to eliminate schema modifications, where possible, especially the ones linked above. If I remember correctly, those were changes that needed to happen after 8.9 IO freeze, but they needed to be changed after Alpha testing where the autogenerated schema was too strict for existing workflows. Transition rules would be the better approach.

**Comment by Myoldmopar** [comment]
This is good stuff.  Thanks for the effort @mjwitte, and I'll just change the milestone to future.

**Comment by mjwitte** [comment]
- [x] Output:Table:SummaryReport names - synonyms need to be transitioned away to match the IDD choices: ABUPS, BEPS, IVRS, CSS, SHAD, and EIO.
*Addressed in #7523.*
